### PR TITLE
Redirect stdout/stderr to Console Widget

### LINKED ISSUE #966: Console output should be in Cutter (Console widget) and not in the executing Terminal
**Environment information**
* Operating System: Any
* Cutter version: GIT Version
* File format: ANY

**Describe the bug**
1. Too much useless (debug?) information is printed to the terminal instead of to Cutter's console widget.
2. Many r2 commands' results are printed to the Terminal and not to Cutter's console widget. 
3. Usually on Windows user executes Cutter by double-clicking the exe file so no Terminal involves. This cause them to miss all the information printed to the terminal.

**To Reproduce**
1. Open Cutter from a terminal `./Cutter"
2. Play with Cutter for awhile
3. Look at the output on the terminal

**Expected behavior**
1. All information should be printed to Cutter console's widget and not to the terminal.
2. There is too many noise created by both r2 and Cutter which is not relevant to the user.

**Screenshots**

Analysis info:
image

Tasks info:
image

Command errors:
image

Command output:
image

r2 noise:
image

### LINKED ISSUE #1434: Redirect STDOUT to Console Widget for Python Plugins
**Is your feature request related to a problem? Please describe.**

Currently, when Plugin author uses `print("hello, world")`, the output would be executed on the terminal which was executing Cutter (on windows usually there won't be a terminal for this).

This is unexpected behavior since we have a Console Widget

**Describe the solution you'd like**
I want STDOUT to be redirected to the Console widget. In general, but specifically here for Python Plugins.

**Additional context**

related issue: #966

### PR DESCRIPTION
**Pull request by yossizap**
**Detailed description**

Currently, python plugin output printed by print() and radare/cutter output printed into stderr/stdout is only shown in console.
This branch adds stdout/stderr redirection to the console widget in addition to the regular console.

**Test plan (required)**

Create a plugin that prints a prepared text with the text formatting available in r2 at a very high rate, compare the console widget output to the text file to make sure there weren't any loses and visually go over the output to confirm that the text was formatted correctly.

**Closing issues**
closes #1434 
closes #966

### CODE REVIEW

**Comment by yossizap** [comment]
This is WIP as I am still working on the Windows solution but I would like to get your opinion on the direction so far.

The main issue here is that radare prints incomplete strings that it updates later with carriage right(For example, "[ ] analysis.." -> "[x] analysis.." which uses https://github.com/radareorg/radare2/blob/master/libr/util/print.c#L2282). 

My solution seems to work without "live" line updates but I am not sure if it's worth investing time into a more complicated solution that will implement carriage right overwrites using cursors.

**Comment by yossizap** [comment]
Extensively tested on Ubuntu 18.04 and Windows 10, everything seems to work well.

I think that terminal emulation should be a separate issue since it's far more complicated to implement while this PR could already be merged.

The changes in the last commit were the most I could do to merge platform specific code.
I tried using _pipe in windows but the file descriptor doesn't support any of the methods [Qt's Windows event dispatcher](https://github.com/qt/qtbase/blob/5.15/src/corelib/kernel/qeventdispatcher_win.cpp) uses to check before emitting readyRead for it to work in QSocketNotifier or QLocalSocket. Using QLocalServer instead of CreateNamedPipe didn't work with the redirection and made the implementation way more complicated, QFile doesn't have any signals and UDP/TCP sockets or new threads seemed like a less favorable option.

**Comment by pelijah** [comment]
Cutter from Appveyor crashes on my Win10.

File: src/widgets/ConsoleWidget.cpp [hunk 1]
```diff
@@ -324,3 +340,57 @@ void ConsoleWidget::invalidateHistoryPosition()
...
+    pipeSocket = new QLocalSocket(this);
+
+    origStderr = fdopen(dup(fileno(stderr)), "a");
+    origStdout = fdopen(dup(fileno(stdout)), "a");
+
+#ifdef Q_OS_WIN
+    QString pipeName = QString::fromLatin1(PIPE_NAME).arg(QUuid::createUuid().toString());
+
+    SECURITY_ATTRIBUTES attributes = {sizeof(SECURITY_ATTRIBUTES), 0, false};
+    hWrite = CreateNamedPipe((wchar_t*)pipeName.utf16(), PIPE_ACCESS_DUPLEX | FILE_FLAG_OVERLAPPED,
```
Comment (pelijah) [inline src/widgets/ConsoleWidget.cpp]: Use `CreateNamedPipeW()` explictily or convert `pipeName` to corresponding type based on `UNICODE` macro's value.

**Comment by yossizap** [comment]
> Cutter from Appveyor crashes on my Win10.

You're right, I am able to reproduce the issue when running the .exe directly, I have only tested in poweshell with
`Start-Process .\cutter.exe -RedirectStandardError stderr.txt -RedirectStandardO
utput stdout.txt`

The crash is caused by [backing up the original stdout and stderr](https://github.com/yossizap/cutter/blob/3ddc0a322f7747049ac4b9897b2d4000e8c2107d/src/widgets/ConsoleWidget.cpp#L362-L363) without the corresponding -Redirect flag for each. After removing both lines and the redirection of stdout to directly print to stdout while redirecting stderr, I discovered that it doesn't work without the -RedirectStandardError flag.

This doesn't seem to happen in a small POC Qt console project I have created, is there something in Cutter's windows build configuration that might be related to this issue that doesn't happen in regular console projects? Also seems to be related to not having stdout/stderr in windows console at all.
I tried AttachConsole and AllocConsole but it didn't help and stdout/stderr weren't printed anyways.

Would appreciate some insight from people who are more familiar with Windows, currently looking into powershell's source code to figure out how those flags influence the program.

**Comment by yossizap** [comment]
This took a while and barely had any documentation/reference for some reason but after a lot of attempts things seem to work well without a console/-RedirectStandardX with zero voodoo.
Also, windows cmd/powershell prints cutter's/r2's output now (with colors+proper formatting).

I couldn't find a clean way to enable output when running the .exe directly, from my research it seems that the only way to do it is to replace AttachConsole with AllocConsole but that will create a cmd window along with cutter, I am not sure if this is desirable so I didn't go that route. Instead, when running the .exe in windows you get a ""Run cutter in a console to enable r2 output redirection into this widget." message in the console widget, making it optional for the user.

**Comment by ITAYC0HEN** [comment]
will review tomorrow on Arch. if you have a sample plugin you wrote for testing, please share the code :)

**Comment by yossizap** [comment]
I had trouble setting up plugins in my build so I just added a timer in ConsoleWidget with random output.

**Comment by ITAYC0HEN** [comment]
Okay tested it, overall it work good :) Especially simple printings from Python plugins I tested with.

here are some feedbacks :)


 - Empty lines in between should be removed (when not expected)

```
[x] Analyze all flags starting with sym. and entry0 (aa)

[x] Analyze function calls (aac)

[x] Analyze len bytes of instructions for references (aar)
```

 - The output is printed both in the Console widget and the Terminal itself. Is it expected?
 - executing `!echo hello` from the console widget will output in the terminal
 -

**Comment by yossizap** [comment]
>Empty lines in between should be removed (when not expected)
Fixed in the latest commit, misunderstood those as intentional.

>The output is printed both in the Console widget and the Terminal itself. Is it expected?
Yes, I intentionally kept the interface as it currently is by printing everything to stderr. This also allows redirection to files and lets the user see what happened if cutter crashes. Could make it toggleable with a flag, up to you.

>executing !echo hello from the console widget will output in the terminal
R2 prints it so it gets redirected to the widget + regular console along with anything else in stdout/stderr. Is this really an issue? Might be useful when running a command that the console widget can't format properly.

**Comment by ITAYC0HEN** [comment]
> Could make it toggleable with a flag, up to you.
@karliss @thestr4ng3r what do you think? Should we show it in the console as well? Should we add a command-line flag to toggle it? Will users expect the Terminal to be empty from noise, or would they prefer to have this info twice? What should be the default state?

> so it gets redirected to the widget + regular console

In reality it does not show up in the Console widget
image

While this shell output does show

```
[0x0040188a]> !lsfsafsa

sh: lsfsafsa: command not found
```

**Comment by yossizap** [comment]
Ok ran it a bunch of times and had missing output as well. From the quick flashing of the output before it's removal it seems to be related to the cursor interaction with executeCommand(also calls removeLastLine()).

linux

I will look into it.

**Comment by ITAYC0HEN** [comment]
```
!!uname -a
```
will print only on Console widget and not on Terminal.

**Comment by yossizap** [comment]
I think that someone should also try out windows to see if they get the OpenType errors that I mentioned

    OpenType support missing for "Segoe UI", script 66
    OpenType support missing for "MS Shell Dlg 2", script 66
    OpenType support missing for "Arial", script 66"

Might be an unrelated issue though

**Pull request merged**